Added -p and -l options to umask, builtin and trap built-ins

JohnoKing opened the pull request:
This pull request adds the following options for printing reinputtable command output:
  `umask -p`    (ksh93v- backport)
  `builtin -p`  (ksh93v- backport)
  ~`hash -l`~

Additionally, `trap -l` has been backported from ksh93v- for listing signals.

JohnoKing committed:
Added -p and -l options to hash, umask, builtin and trap built-ins

The following options have been added for printing reinputtable
command output:
  umask -p    (ksh93v- backport)
  builtin -p  (ksh93v- backport)
  hash -l

Additionally, 'trap -l' has been backported from ksh93v- for listing
signals.

McDutchie commented on src/cmd/ksh93/data/builtins.c:
```diff
@@ -1502,7 +1508,7 @@ const char sh_optreadonly[] =
 	"field splitting and pathname expansion are not performed on "
 	"the arguments. Tilde expansion occurs on \avalue\a.]"
-"[p?Causes the output to be in a form of \breadonly\b commands that can be "
+"[p?Causes the output to be in the form of \breadonly\b commands that can be "
```
I think this should stay the way it is. There are multiple such forms: assignment decoupled from `readonly` commands, or `readonly` with assignment but different quoting styles.

JohnoKing committed:
Revert to 'a form of'

McDutchie committed:
Merge branch 'dev' into builtin-reinput

McDutchie commented:
Did you have any use case in mind for `hash -l`? Since tracked aliases are considered obsolete anyway, and they don't really affect command execution at all (any invalid entry will be ignored), I'm not really seeing it.

JohnoKing committed:
Fix builtin -p bug backported from 93v-

JohnoKing commented:
I've fixed `builtin -p` to error out if the provided builtins don't exist. Regarding `/opt/ast/bin` builtins, it will not register the builtin's existence unless it's first loaded or if it's accessed by the direct virtual path:
```
$ builtin -p cat
builtin: cat: not found
$ builtin cat
$ builtin -p cat
builtin cat
$ builtin -p /opt/ast/bin/cut cut
builtin /opt/ast/bin/cut
builtin: cut: not found
$ builtin cut
$ builtin -p /opt/ast/bin/cut cut
builtin /opt/ast/bin/cut
builtin cut
```

> Did you have any use case in mind for `hash -l`?

Not particularly. The main reason I added it was because bash has it, and because the feature was cheap to implement as far as LOC goes.

JohnoKing committed:
Merge branch 'dev' into builtin-reinput

JohnoKing committed:
Merge branch 'dev' into builtin-reinput

McDutchie commented:
> Not particularly. The main reason I added it was because bash has it, and because the feature was cheap to implement as far as LOC goes.
> 
> _Edit 06-09-2025:_ I'l also note that the `-l` option gives the `hash` builtin feature parity with tracked alias listing obtainable via the already implemented `alias -pt`.



> I've fixed `builtin -p` to error out if the provided builtins don't exist. Regarding `/opt/ast/bin` builtins, it will not register the builtin's existence unless it's first loaded or if it's accessed by the direct virtual path:
> 
> ```
> $ builtin -p cat
> builtin: cat: not found
> $ builtin cat
> $ builtin -p cat
> builtin cat
> $ builtin -p /opt/ast/bin/cut cut
> builtin /opt/ast/bin/cut
> builtin: cut: not found
> $ builtin cut
> $ builtin -p /opt/ast/bin/cut cut
> builtin /opt/ast/bin/cut
> builtin cut
> ```
> 
> > Did you have any use case in mind for `hash -l`?
> 
> Not particularly. The main reason I added it was because bash has it, and because the feature was cheap to implement as far as LOC goes.
> 
> _Edit 06-09-2025:_ I'l also note that the `-l` option gives the `hash` builtin feature parity with tracked alias listing obtainable via the already implemented `alias -pt`.

However, the bash feature is different:

```
$ bash
bash-5.2$ ls >/dev/null
bash-5.2$ hash -l
builtin hash -p /bin/ls ls
```

So bash produces commands that recreate the tracked aliases as they were, without searching them in PATH again. This is the true meaning of "recreating the current tracked aliases". Your ksh version doesn't do this. Even if it did, it wouldn't make a difference, because unlike bash, ksh effectively ignores the hash table when doing PATH searches.

So I don't think `hash -l` is useful, and I'd rather have this PR without it.

JohnoKing committed:
Merge remote-tracking branch 'upstream/dev' into builtin-reinput

JohnoKing committed:
rm rejected hash -l

McDutchie committed:
Merge branch 'dev' into builtin-reinput

McDutchie committed:
update ANNOUNCE

McDutchie committed:
update version dates and copyright years

McDutchie committed:
undo unrelated/cosmetic diffs (these are for cleanups commits)
